zio-http/src/main/scala/zhttp/socket/SocketDecoder.scala

```diff
@@ -5,77 +5,69 @@ import io.netty.handler.codec.http.websocketx.WebSocketDecoderConfig
-        case SkipUTF8Validation            => b.withUTF8Validator(false)
-        case Concat(a, b)                  =>
-          loop(a)
-          loop(b)
-      }
-      ()
-    }
-    loop(self)
-    b.build()
-  }
-}
-
-object SocketDecoder {
-  private final case class MaxFramePayloadLength(length: Int)         extends SocketDecoder
-  private case object RejectMaskedFrames                              extends SocketDecoder
-  private case object AllowMaskMismatch                               extends SocketDecoder
-  private case object AllowExtensions                                 extends SocketDecoder
-  private case object AllowProtocolViolation                          extends SocketDecoder
-  private case object SkipUTF8Validation                              extends SocketDecoder
-  private final case class Concat(a: SocketDecoder, b: SocketDecoder) extends SocketDecoder
-  private case object Default                                         extends SocketDecoder
+final case class SocketDecoder(
+  maxFramePayloadLength: Int = 65536,
+  expectMaskedFrames: Boolean = true,
+  allowMaskMismatch: Boolean = false,
+  allowExtensions: Boolean = false,
+  closeOnProtocolViolation: Boolean = true,
+  withUTF8Validator: Boolean = true,
+) { self =>
 
   /**
    * Sets Maximum length of a frame's payload. Setting this to an appropriate
    * value for you application helps check for denial of services attacks.
    */
-  def maxFramePayloadLength(length: Int): SocketDecoder = MaxFramePayloadLength(length)
+  def withMaxFramePayloadLength(length: Int): SocketDecoder = self.copy(maxFramePayloadLength = length)
 
   /**
    * Web socket servers must set this to true to reject incoming masked payload.
    */
-  def rejectMaskedFrames: SocketDecoder = RejectMaskedFrames
+  def withRejectMaskedFrames: SocketDecoder = self.copy(expectMaskedFrames = true)
+
+  def withAllowMaskedFrames: SocketDecoder = self.copy(expectMaskedFrames = false)
 
   /**
    * When set to true, frames which are not masked properly according to the
    * standard will still be accepted.
    */
-  def allowMaskMismatch: SocketDecoder = AllowMaskMismatch
+  def withAllowMaskMismatch: SocketDecoder = self.copy(allowMaskMismatch = true)
 
-  /**
-   * Allow extensions to be used in the reserved bits of the web socket frame
-   */
-  def allowExtensions: SocketDecoder = AllowExtensions
+  def withRejectMaskMismatch: SocketDecoder = self.copy(allowMaskMismatch = false)
+
+  def withAllowExtensions: SocketDecoder  = self.copy(allowExtensions = true)
+  def withRejectExtensions: SocketDecoder = self.copy(allowExtensions = false)
 
   /**
    * Flag to not send close frame immediately on any protocol violation.ion.
    */
-  def allowProtocolViolation: SocketDecoder = AllowProtocolViolation
+  def withAllowProtocolViolation: SocketDecoder = self.copy(closeOnProtocolViolation = false)
+
+  def withRejectProtocolViolation: SocketDecoder = self.copy(closeOnProtocolViolation = true)
```

Comment:
```scala
def withMaskMismatch(allowed: Boolean): SocketDecoder = ...
def withProtocolVoilation(allowed: Boolean): SocketDecoder = ...
```

I think this reduces the surface areas of the API, which longterm reduces the probability of producing a breaking change.